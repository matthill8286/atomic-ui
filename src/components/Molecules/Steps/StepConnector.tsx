import React, { FC } from 'react'
import { css, styled } from '@/styles/styled'
import {
  StepConnectorLineProps,
  StepConnectorPositions,
  StepConnectorProps,
} from './Steps.interface'

const getStepConnectorPositions = ({
  theme,
  variant,
  elementPosition,
  rightMargin,
  leftMargin,
}): StepConnectorPositions => {
  const positions: StepConnectorPositions = {
    uniform: {
      first: css`
        left: 50%;
        right: 0;
      `,
      second: css`
        left: -50%;
        right: 50%;
      `,
      last: css`
        left: -50%;
        right: 50%;
      `,
      default: css`
        left: -50%;
        right: 50%;
      `,
    },
    prominent: {
      first: css`
        left: 10px;
        right: calc(100% - ${theme.spacing.base.md});
      `,
      second: css`
        left: calc(-100% + ${theme.spacing.base.md});
        right: 50%;
      `,
      last: css`
        left: -50%;
        right: 10px;
      `,
      default: css`
        left: -50%;
        right: 50%;
      `,
    },
    simple: {
      first: css`
        display: none;
      `,
      second: css`
        left: calc(-100% + ${42 + leftMargin}px);
        right: calc(50% + ${28 + rightMargin}px);
      `,
      last: css`
        left: calc(-50% + ${28 + leftMargin}px);
        right: calc(0% + ${42 + rightMargin}px);
      `,
      default: css`
        left: calc(-50% + ${28 + leftMargin}px);
        right: calc(50% + ${28 + rightMargin}px);
      `,
    },
  }

  return positions[variant][elementPosition]
}

const BottomPositions = {
  uniform: {
    completed: '16px',
    default: '8px',
  },
  prominent: {
    completed: '16px',
    default: '8px',
  },
  simple: {
    completed: '5px',
    default: '5px',
  },
}

export const StepConnectorRoot = styled.div<StepConnectorProps>(
  ({
    theme,
    elementPosition = 'default',
    active,
    completed,
    prevActive,
    variant = 'prominent',
  }) => {
    const leftMargin = active ? 4 : 0
    const rightMargin = prevActive ? 4 : 0

    const currentPosition = getStepConnectorPositions({
      theme,
      variant,
      elementPosition,
      leftMargin,
      rightMargin,
    })

    const bottomPosition = completed
      ? BottomPositions[variant]['completed']
      : BottomPositions[variant]['default']

    return css`
      position: absolute;
      bottom: ${bottomPosition};
      ${currentPosition};
    `
  }
)

export const StepConnectorLine = styled.span<StepConnectorLineProps>`
  border-color: ${({ theme, color = 'primary', active, completed, elementPosition = 'default' }) =>
    completed || active || elementPosition === 'first' ? theme.color[color] : theme.color.grey2};
  border-style: solid;
  border-top-width: ${({ borderWidth }) => borderWidth};
  border-bottom-width: ${({ borderWidth }) => borderWidth};
  border-radius: ${({ theme }) => theme.dimension.borderRadius1};
  display: block;
`

export const StepConnector: FC<StepConnectorProps> = props => (
  <StepConnectorRoot {...props}>
    <StepConnectorLine {...props} borderWidth={props.variant === 'simple' ? '2px' : '4px'} />
  </StepConnectorRoot>
)

StepConnector.displayName = 'StepConnector'
