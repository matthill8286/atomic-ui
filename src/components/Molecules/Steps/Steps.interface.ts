import { ReactElement } from 'react'
import { Elevation, ThemeColors } from '@/types/theme'

export type StepsVariant = 'prominent' | 'simple' | 'uniform'

export interface StepConnectorProps {
  color?: ThemeColors
  active?: boolean
  prevActive?: boolean
  completed?: boolean
  elementPosition?: ElementPosition
  variant?: StepsVariant
}

export interface StepConnectorLineProps extends StepConnectorProps {
  borderWidth: string
}

export interface CustomIconProps {
  active?: boolean
  completed?: boolean
  color?: ThemeColors
}

export interface StepIconProps {
  active?: boolean
  completed?: boolean
  stepNumber?: string
  color?: ThemeColors
  variant?: StepsVariant
  customIcon?: ReactElement<{ active: boolean; completed: boolean; color: ThemeColors }>
}

export type ElementPosition = 'default' | 'first' | 'second' | 'last'

export type StepConnectorPositions = {
  [key in StepsVariant]: {
    // eslint-disable-next-line @typescript-eslint/ban-types
    [key in ElementPosition]: {}
  }
}

export interface StepPublicProps {
  title?: string
  description?: string
}

export interface StepPrivateProps {
  color?: ThemeColors
  active?: boolean
  completed?: boolean
  elementPosition?: ElementPosition
  stepIndex?: number
  icon?: ReactElement<{ active: boolean; completed: boolean; color: ThemeColors }>
  connector?: ReactElement<StepConnectorProps>
  variant?: StepsVariant
  onClick?: (stepIndex: number) => void
}

export interface StepsProps {
  activeStep: number
  completedStep: number
  elevation?: Elevation
  color?: ThemeColors
  children?: ReactElement<StepPublicProps>[]
  connector?: ReactElement<StepConnectorProps>
  variant?: StepsVariant
  onChange?: (step: number) => void
}
