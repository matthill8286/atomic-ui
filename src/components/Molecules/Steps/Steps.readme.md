## Steps

Steps display progress through numbered steps. It provides a wizard-like workflow.

#### API

To control the `Steps` we can se the `activeStep` and the `completedStep`.

There are 2 variants: `prominent` and `simple`.

It follows the `ThemeColors` with the `color` prop.

`onChange` prop will be called when a Step is clicked and it will get the index as argoument.

#### Step

We need to pass `title` and `description?` props, all the other props of the `Step` are implicitly passed from the `Steps` component.

We can change the connector component, a custom StepConnector component can be passed to the `Steps.connector` prop.

The `Step` component will receive automatically the `CustomStepConnector` from the Steps component.

Description hides when `media.maxMd`.
