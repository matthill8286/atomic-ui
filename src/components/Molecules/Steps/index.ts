export * from './StepConnector'
export * from './Step'
export * from './Steps'
export * from './Steps.interface'
