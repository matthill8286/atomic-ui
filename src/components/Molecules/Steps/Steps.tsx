import React, { Children, cloneElement, FC, isValidElement } from 'react'
import { Card } from '@/components/Atoms/Card'
import { FlexBox } from '@/components/Helper/FlexBox'
import { StepConnector } from './StepConnector'
import {
  ElementPosition,
  StepConnectorProps,
  StepPrivateProps,
  StepPublicProps,
  StepsProps,
} from './Steps.interface'

export const Steps: FC<StepsProps> = props => {
  const {
    activeStep = 0,
    completedStep = 0,
    elevation = 0,
    connector,
    children,
    color = 'primary',
    variant = 'prominent',
    onChange,
    ...other
  } = props

  const childrenArray = Children.toArray(children)

  const steps = childrenArray.map((step, stepIndex) => {
    if (isValidElement(step)) {
      const active = activeStep === stepIndex
      const completed = stepIndex <= completedStep

      let elementPosition: ElementPosition = 'default'

      if (stepIndex === 0) elementPosition = 'first'
      if (stepIndex === 1) elementPosition = 'second'
      if (stepIndex + 1 === Children.count(children)) elementPosition = 'last'

      const connectorProps: StepConnectorProps = {
        elementPosition,
        active: stepIndex <= activeStep,
        prevActive: stepIndex - 1 === activeStep,
        completed,
        color,
        variant,
      }

      const connectorComponent = isValidElement(connector) ? (
        cloneElement(connector, { ...connectorProps, ...connector.props })
      ) : (
        <StepConnector {...connectorProps} />
      )

      const stepProps: StepPublicProps & StepPrivateProps = {
        ...step.props,
        connector: connectorComponent,
        elementPosition,
        stepIndex,
        active,
        completed,
        color,
        onClick: onChange,
        variant,
      }

      return cloneElement(step, stepProps)
    }
  })

  return (
    <Card
      elevation={elevation}
      noBorder={['top', 'right', 'bottom', 'left']}
      surface="clear"
      padding="md"
      {...other}>
      <FlexBox display="flex" alignItems="center" flexDirection="row">
        {steps}
      </FlexBox>
    </Card>
  )
}

Steps.displayName = 'Steps'
