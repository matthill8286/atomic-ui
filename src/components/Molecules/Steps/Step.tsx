import React, { FC } from 'react'
import { Typo } from '@/components/Atoms/Typography'
import { media } from '@/styles/media'
import { styled } from '@/styles/styled'
import { ThemeColors } from '@/types/theme'
import { StepIcon } from './StepIcon'
import {
  ElementPosition,
  StepIconProps,
  StepPrivateProps,
  StepPublicProps,
} from './Steps.interface'

const StepRoot = styled.div<{ pointer: boolean }>`
  cursor: ${({ pointer }) => (pointer ? 'pointer' : 'default')};
  flex: 1;
  position: relative;
  padding: 0 ${({ theme }) => theme.defaultSpacing}px;
`

const StepLabelRoot = styled.span<{ elementPosition: ElementPosition }>`
  flex-direction: column;
  display: flex;
  align-items: center;

  ${({ elementPosition }) => elementPosition === 'first' && 'align-items: flex-start;'}
  ${({ elementPosition }) => elementPosition === 'last' && 'align-items: flex-end;'}
`

const StepDescription = styled(Typo)`
  ${media.maxMd} {
    display: none;
  }
`
export const Step: FC<StepPublicProps & StepPrivateProps> = props => {
  const {
    color = 'primary',
    elementPosition = 'default',
    active = false,
    completed = false,
    stepIndex = 0,
    title = '',
    description = '',
    icon,
    connector,
    onClick,
    variant = 'prominent',
    ...other
  } = props

  const stepIconProps: StepIconProps = {
    active,
    completed,
    color,
    variant,
    customIcon: icon,
  }

  const textColor: ThemeColors = active || completed ? color : 'grey2'

  return (
    <StepRoot {...other} pointer={!!onClick} onClick={() => onClick && onClick(stepIndex)}>
      {connector}
      {variant != 'uniform' ? (
        <StepLabelRoot elementPosition={elementPosition}>
          {title && (
            <Typo
              tag="span"
              color={textColor}
              weight={active || completed ? 'semibold' : 'regular'}>
              {title}
            </Typo>
          )}
          {description && (
            <StepDescription tag="span" color={textColor}>
              {description}
            </StepDescription>
          )}
          <StepIcon {...stepIconProps} />
        </StepLabelRoot>
      ) : (
        <StepIcon {...stepIconProps} />
      )}
    </StepRoot>
  )
}

Step.displayName = 'Step'
