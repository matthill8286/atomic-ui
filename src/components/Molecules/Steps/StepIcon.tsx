import React, { cloneElement, FC, isValidElement } from 'react'
import { Typo } from '@/components/Atoms/Typography'
import { css, styled } from '@/styles/styled'
import { ThemeColors } from '@/types/theme'
import { StepIconProps, StepsVariant } from './Steps.interface'

const FlexCenter = css`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
`

const StepIconRoot = styled.div<{ variant: StepsVariant; active: boolean }>`
  margin-top: ${({ theme, variant, active }) =>
    variant === 'simple' && !active ? '12px' : `${theme.defaultSpacing}px`};
  margin-bottom: ${({ variant, active }) => (variant === 'simple' && !active ? '2px' : '0')};
  position: relative;
  z-index: 1;
  ${FlexCenter}
`

const Circle = styled.div<{ active: boolean; completed: boolean; color: ThemeColors }>`
  border: ${({ active, completed }) => (completed ? '6px' : active ? '4px' : '4px')} solid
    ${({ theme, active, completed, color }) =>
      theme.color[active ? color : completed ? 'white' : `grey2`]};
  border-radius: 50%;
  width: ${({ completed }) => (completed ? '40px' : '24px')};
  height: ${({ completed }) => (completed ? '40px' : '24px')};
  background-color: ${({ theme, active, completed, color }) =>
    theme.color[active ? `white` : completed ? color : 'grey2']};
  box-shadow: ${({ active, completed }) => (active || completed ? '2px 2px 4px' : '0 0 0')}
    ${({ theme }) => theme.color['grey2']};
  ${FlexCenter}
`

const Checkmark = styled.div<{ active: boolean; color: ThemeColors }>`
  position: relative;
  display: inline-block;
  width: ${({ theme }) => theme.spacing.base.lg};
  height: ${({ theme }) => theme.spacing.base.lg};

  ::before {
    position: absolute;
    left: 0;
    top: 13px;
    height: 31%;
    width: 2px;
    background-color: ${({ theme }) => theme.color['white']};
    content: '';
    transform: translateX(12px) rotate(-45deg);
    transform-origin: left bottom;
  }

  ::after {
    position: absolute;
    left: 0;
    bottom: 9px;
    height: 2px;
    width: 55%;
    background-color: ${({ theme }) => theme.color['white']};
    content: '';
    transform: translateX(12px) rotate(-45deg);
    transform-origin: left bottom;
  }
`

const Point = styled.div<{ backgroundColor: ThemeColors; active: boolean }>(
  ({ theme, active, backgroundColor }) => ({
    border: `2px solid ${theme.color[backgroundColor]}`,
    backgroundColor: theme.color[backgroundColor],
    borderRadius: '50%',
    width: active ? 12 : 6,
    height: active ? 12 : 6,
  })
)

export const StepIcon: FC<StepIconProps> = props => {
  const {
    active = false,
    completed = false,
    color = 'primary',
    variant = 'prominent',
    customIcon,
    ...other
  } = props

  const renderIcon = () => {
    if (customIcon && isValidElement(customIcon)) {
      return (
        <Circle active={active} completed={completed} color={color}>
          {cloneElement(customIcon, { active, completed, color })}
        </Circle>
      )
    }

    if (variant === 'simple') {
      return <Point backgroundColor={active || completed ? color : 'grey2'} active={active} />
    }

    return (
      <Circle active={active} completed={completed} color={color}>
        {completed ? (
          <Checkmark active={active} color={color} />
        ) : (
          <Typo color={active ? 'white' : color}></Typo>
        )}
      </Circle>
    )
  }

  return (
    <StepIconRoot variant={variant} active={active} {...other}>
      {renderIcon()}
    </StepIconRoot>
  )
}

StepIcon.displayName = 'StepIcon'
