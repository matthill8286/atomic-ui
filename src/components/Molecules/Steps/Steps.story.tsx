import { withState } from '@dump247/storybook-state'
import { button, radios, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React, { FC } from 'react'
import { Icon } from '@/components/Atoms/Icon'
import { styled } from '@/styles/styled'
import { IconArticle, IconCheckmarkCircleOutlined } from '@matthill8286/jsx-icon-library'
import { ThemeColors } from '@/types/theme'
import { Step } from './Step'
import { StepConnectorRoot } from './StepConnector'
import { Steps } from './Steps'
import {
  CustomIconProps,
  StepConnectorLineProps,
  StepConnectorProps,
  StepsProps,
  StepsVariant,
} from './Steps.interface'
import readme from './Steps.readme.md'
import withVariableArrayReadme from './StepsVariableArray.readme.md'
import withConnectorReadme from './StepsWithConnector.readme.md'
import withIconReadme from './StepsWithIcon.readme.md'

const stories = storiesOf('Design System/Molecules/Steps', module)

const variants: { [key: string]: StepsVariant } = {
  prominent: 'prominent',
  uniform: 'uniform',
  simple: 'simple',
}

const colors: ThemeColors[] = ['primary', 'secondary', 'black']

stories.add(
  'Default',
  () => {
    const steps = [{ title: 'One' }, { title: 'Two' }, { title: 'Three' }, { title: 'Four' }]
    const stepsIndex: number[] = steps.map((el, index) => index)

    const knobs: StepsProps = {
      color: select('color', colors, 'primary'),
      activeStep: select('activeStep', stepsIndex, 2),
      completedStep: select('completedStep', [-1, ...stepsIndex], 1),
      variant: radios('variant', variants, 'uniform'),
    }

    const description = text('description', 'This is a description')

    return (
      <Steps {...knobs}>
        {steps.map(({ title }, index) => (
          <Step key={index} title={title} description={description} />
        ))}
      </Steps>
    )
  },
  {
    info: readme,
  }
)

stories.add(
  'onChange activeStep',
  withState({ activeStep: 2, completedStep: 1 }, store => {
    const stepsProps: StepsProps = {
      color: 'primary',
      activeStep: store.state.activeStep,
      completedStep: store.state.completedStep,
    }

    const handleChange = (step: number) => {
      store.set({ activeStep: step, completedStep: step - 1 })
    }

    return (
      <Steps {...stepsProps} onChange={handleChange}>
        <Step title="Sign In" description="This is a description" />
        <Step title="Address" description="This is a description" />
        <Step title="Payment" description="This is a description" />
        <Step title="Summary" description="This is a description" />
      </Steps>
    )
  }),
  {
    info: readme,
  }
)

stories.add(
  'with icon',
  () => {
    const steps = [
      { title: 'Sig In' },
      { title: 'Address' },
      { title: 'Payment' },
      { title: 'Summary' },
    ]
    const stepsIndex: number[] = steps.map((el, index) => index)

    const knobs: StepsProps = {
      color: select('color', colors, 'primary'),
      activeStep: select('activeStep', stepsIndex, 2),
      completedStep: select('completedStep', [-1, ...stepsIndex], 1),
    }

    const description = text('description', 'This is a description')

    const CustomIcon: FC<CustomIconProps> = ({
      active = false,
      completed = false,
      color = 'primary',
      ...props
    }) => (
      <Icon color={active ? 'white' : completed ? color : 'grey2'} {...props}>
        <IconCheckmarkCircleOutlined />
      </Icon>
    )

    return (
      <Steps {...knobs}>
        {steps.map(({ title }, index) => (
          <Step key={index} title={title} description={description} icon={<CustomIcon />} />
        ))}
      </Steps>
    )
  },
  {
    info: withIconReadme,
  }
)

stories.add(
  'with connector',
  () => {
    const steps = [
      { title: 'Sig In' },
      { title: 'Address' },
      { title: 'Payment' },
      { title: 'Summary' },
    ]
    const stepsIndex: number[] = steps.map((el, index) => index)

    const knobs: StepsProps = {
      color: select('color', colors, 'primary'),
      activeStep: select('activeStep', stepsIndex, 2),
      completedStep: select('completedStep', [-1, ...stepsIndex], 1),
    }

    const description = text('description', 'This is a description')

    const CustomConnectorLine = styled.span<StepConnectorLineProps>`
      border-color: ${({
        theme,
        color = 'primary',
        active,
        completed,
        elementPosition = 'default',
      }) =>
        completed || active || elementPosition === 'first'
          ? theme.color[color]
          : theme.color.grey2};
      border-style: solid;
      border-top-width: ${({ borderWidth }) => borderWidth};
      border-bottom-width: ${({ borderWidth }) => borderWidth};
      border-radius: ${({ theme }) => theme.dimension.borderRadius1};
      display: block;
    `

    const CustomConnector: FC<StepConnectorProps> = props => (
      <StepConnectorRoot {...props}>
        <CustomConnectorLine {...props} borderWidth={props.variant === 'simple' ? '2px' : '4px'} />
      </StepConnectorRoot>
    )

    return (
      <Steps {...knobs} connector={<CustomConnector color="black" />}>
        {steps.map(({ title }, index) => (
          <Step key={index} title={title} description={description} />
        ))}
      </Steps>
    )
  },
  {
    info: withConnectorReadme,
  }
)

stories.add(
  'Uniform',
  () => {
    const steps = [{ title: 'One' }, { title: 'Two' }, { title: 'Three' }, { title: 'Four' }]
    const stepsIndex: number[] = steps.map((el, index) => index)

    const knobs: StepsProps = {
      color: select('color', colors, 'primary'),
      activeStep: select('activeStep', stepsIndex, 2),
      completedStep: select('completedStep', [-1, ...stepsIndex], 1),
      variant: radios('variant', variants, 'uniform'),
    }

    const description = text('description', 'This is a description')

    return (
      <Steps {...knobs}>
        {steps.map(({ title }, index) => (
          <Step key={index} title={title} description={description} />
        ))}
      </Steps>
    )
  },
  {
    info: readme,
  }
)

stories.add(
  'variable arrays',
  withState(
    {
      steps: [
        { title: 'Sign In' },
        { title: 'Address' },
        { title: 'Payment' },
        { title: 'Summary' },
      ],
      activeStep: 2,
      completedStep: 1,
    },
    store => {
      const {
        state: { steps, activeStep, completedStep },
      } = store

      const knobs: StepsProps = {
        activeStep,
        completedStep,
        color: select('color', colors, 'primary'),
        variant: radios('variant', variants, 'prominent'),
      }

      const increment = button('Add an element', () => {
        store.set({
          steps: [...steps, { title: 'Added' }],
        })
      })

      const decrement = button('Remove an element', () => {
        if (steps.length > 3) {
          store.set({ steps: steps.slice(0, -1) })
        } else {
          store.set({ steps })
        }
      })

      const reset = button('Reset to initial state', () => {
        store.reset()
      })

      const handleChange = (step: number) => {
        store.set({ activeStep: step, completedStep: step - 1 })
      }

      return (
        <Steps {...knobs} onChange={handleChange}>
          {steps.map(({ title }, index) => (
            <Step key={index} />
          ))}
        </Steps>
      )
    }
  ),
  {
    info: withVariableArrayReadme,
  }
)
