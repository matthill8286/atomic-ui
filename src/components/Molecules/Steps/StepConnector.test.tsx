import React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { StepConnector } from './StepConnector'
import { StepConnectorProps } from './Steps.interface'

describe('StepConnector', () => {
  const props: StepConnectorProps = {
    active: false,
    prevActive: false,
    color: 'primary',
    variant: 'prominent',
    completed: false,
    elementPosition: 'first',
  }

  it('should render', () => {
    const tree = renderWithTheme(<StepConnector {...props} />)
    expect(tree).toMatchSnapshot()
  })
})
