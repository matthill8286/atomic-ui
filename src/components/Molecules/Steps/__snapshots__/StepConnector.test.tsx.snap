// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepConnector should render 1`] = `
.c0 {
  position: absolute;
  bottom: 8px;
  left: 10px;
  right: calc(100% - 24px);
}

.c1 {
  border-color: #00C8DC;
  border-style: solid;
  border-top-width: 4px;
  border-bottom-width: 4px;
  border-radius: 4px;
  display: block;
}

<div
  className="c0"
  color="primary"
>
  <span
    className="c1"
    color="primary"
  />
</div>
`;
